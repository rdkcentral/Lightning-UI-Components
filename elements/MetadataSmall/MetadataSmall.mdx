import { Canvas, Story } from '@storybook/addon-docs/blocks';
import { Theme } from '../../.storybook/utils';
import MetadataSmall from '.';

# MetadataSmall

Content to be displayed inside Tiles using the inset metadata position in [withMetadata](?path=/docs/patterns-withmetadata--basic). Displays title, info, progress bar, and logo.

## Usage

Wrap a Tile in the `withMetadata` mixin, and pass `MetadataSmall` as the `Metadata` prop:

```js
import { MetadataSmall, Tile } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      TileWithMetadata: {
        type: withMetadata(Tile),
        src: 'asset/path/pets.png',
        w: 410,
        h: 230,
        color: 0xff5d7a74,
        metadataLocation: 'inset',
        Metadata: {
          type: MetadataSmall,
          title: 'Movie Title',
          data: 'A description of the movie',
          logo: 'url/path/logo.png'
        }
      }
    };
  }
}
```

Include previews:

<Canvas>
  <Story id="patterns-withmetadata--inset-metadata" />
</Canvas>

The `id` for the preview comes from the url path generated for the story.

## API

### Properties

| name     | type                                                     | required | default | description                                                          |
| -------- | -------------------------------------------------------- | -------- | ------- | -------------------------------------------------------------------- |
| title    | string                                                   | false    | -       | the first line of text                                               |
| data     | [InlineContent](?path=/docs/layout-inlinecontent--basic) | false    | -       | relevant content data at the bottom                                  |
| progress | number                                                   | false    | 0       | percentage of the current progress from 0 to 1. Hidden if value is 0 |
| logo     | string                                                   | false    | -       | URI of a logo to display at bottom of component                      |

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={MetadataSmall.styles} />
