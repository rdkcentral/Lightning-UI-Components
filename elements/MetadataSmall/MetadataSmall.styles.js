export default theme => ({
  logoSpacing: theme.spacing(4),
  logoW: 32,
  fadeWidth: 100,
  w: 410,
  justify: 'flex-start',
  marqueeProperties: {},
  firstLineTextProperties: {
    ...theme.typography.headline3,
    textColor: theme.palette.text.light.primary,
    maxLinesSuffix: '...',
    maxLines: 2
  },
  secondLineTextProperties: {
    ...theme.typography.body3,
    textColor: theme.palette.text.light.secondary,
    maxLinesSuffix: '...',
    maxLines: 1
  },
  secondLinePaddingTop: -theme.spacing(1)
});
