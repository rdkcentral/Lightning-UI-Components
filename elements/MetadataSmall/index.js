import Base from '../Base';
import styles from './MetadataSmall.styles';
import withStyles from '../../mixins/withStyles';
import InlineContent from '../../layout/InlineContent';
import Icon from '../Icon';
import { FadeShader } from '../../textures';
import TextBox from '../TextBox';

export default class MetadataSmall extends withStyles(Base, styles) {
  static _template() {
    return {
      Text: {
        flex: { direction: 'column' },
        FirstLine: {
          type: TextBox,
          signals: {
            textBoxChanged: '_textBoxChanged'
          }
        },
        SecondLineWrapper: {
          SecondLine: {
            type: InlineContent,
            y: this.styles.secondLinePaddingTop
          }
        }
      },
      Logo: {
        type: Icon,
        alpha: 0.001
      }
    };
  }

  static get properties() {
    return [
      'logo',
      'logoW',
      'logoH',
      'logoSpacing',
      'firstLine',
      'secondLine',
      'justify',
      'fadeWidth',
      'firstLineTextProperties',
      'secondLineTextProperties',
      'focusScaleConst',
      'unfocusScaleConst'
    ];
  }

  static get tags() {
    return [
      'Text',
      'SecondLineWrapper',
      'Logo',
      'FirstLine',
      { name: 'SecondLine', path: 'SecondLineWrapper.SecondLine' }
    ];
  }

  _construct() {
    super._construct();
    this._logoSpacing = this.styles.logoSpacing;
    this._logoW = this.styles.logoW;
    this._justify = this.styles.justify;
    this._firstLineTextProperties = this.styles.firstLineTextProperties;
    this._secondLineTextProperties = this.styles.secondLineTextProperties;
    this._fadeWidth = this.styles.fadeWidth;
    this._w = this.styles.w;
    this._secondLinePaddingTop = this.styles.secondLinePaddingTop;
  }

  _init() {
    if (this.focusScaleConst === undefined) {
      this._focusScaleConst = this._getFocusScale(this.w, this.h);
    }
    if (this.unfocusScaleConst === undefined) {
      this._unfocusScaleConst = this._getUnfocusScale(this.w, this.h);
    }
    this._enabledLogoListeners();
    super._init();
  }

  _enabledLogoListeners() {
    this._Logo.on('txLoaded', () => this._logoTxListenerCb(false));
    this._Logo.on('txError', () => this._logoTxListenerCb(true));
  }

  _logoTxListenerCb(isError = false) {
    this.logo = isError ? null : this.logo;
    this._logoLoaded = !isError;
    this._requestUpdateDebounce();
    this._alphaLogo();
  }

  _alphaLogo() {
    const alpha = this.logo ? 1 : 0;
    if (this._smooth) {
      this._Logo.smooth = { alpha };
    } else {
      this._Logo.patch({ alpha });
    }
  }

  $loadedInlineContent(line) {
    if (line.ref && this.tag(line.ref + 'Wrapper')) {
      this.tag(line.ref + 'Wrapper').h = line.textHeight;
      this.tag(line.ref + 'Wrapper').alpha = 1;
    }
    line.h = line.content ? line.textHeight : 0;
    this._updateComponentDimensions();
  }

  _textBoxChanged() {
    this._updateComponentDimensions();
  }

  _update() {
    this._updateFirstLine();
    this._updateSecondLine();
    this._updateLogo();
  }

  _updateComponentDimensions() {
    this._Text.h = this._FirstLine.renderHeight + this._SecondLine.h;
    this.h = this._FirstLine.renderHeight + this._SecondLine.renderHeight;
    this._updateLogo();
  }

  _updateLogo() {
    this._Logo.w = this.logoW;
    if (this.logoH) {
      this._Logo.h = this.logoH;
    }

    const logoH = this.logoH || this._Logo.finalH || this.h;
    const logoX = this._textW + this.logoSpacing;

    const logoY =
      this._Text.h > logoH
        ? (this._Text.h - logoH) / 2
        : (logoH - this._Text.h) / 2;

    if (this._smooth && this._logoLoaded) {
      this._Logo.smooth = { x: logoX, y: logoY };
    } else {
      this._Logo.patch({ x: logoX, y: logoY });
    }

    if (this._Logo.icon !== this.logo) {
      this._logoLoaded = false;
      this._Logo.icon = this.logo;
    }
  }

  _updateFirstLine() {
    this._firstLineObject.component.patch({
      content: this._firstLineObject.content,
      wordWrap: true,
      wordWrapWidth: this._textW,
      style: this._firstLineTextProperties
    });
  }

  _updateSecondLine() {
    this._secondLineObject.component.patch({
      justify: this._justify,
      content: this._secondLineObject.content,
      textProperties: this._secondLineObject.textProps
    });
    this._updateShader(this._secondLineObject);
    this._secondLineObject.wrapper.visible = this._secondLineObject.content
      ? true
      : false;
  }

  _updateShader(line) {
    if (line.component.finalW > this._textW) {
      line.wrapper.patch({
        w: this._textW + this.fadeWidth / 2,
        shader: {
          type: FadeShader,
          positionLeft: 0,
          positionRight: this.fadeWidth
        },
        rtt: true
      });
    } else {
      line.wrapper.shader = undefined;
    }
  }

  get _linesArray() {
    return [this._firstLineObject, this._secondLineObject];
  }

  get _firstLineObject() {
    return {
      component: this._FirstLine,
      content: this._firstLine,
      textProps: this._firstLineTextProperties
    };
  }

  get _secondLineObject() {
    return {
      wrapper: this._SecondLineWrapper,
      component: this._SecondLine,
      content: this._secondLine,
      textProps: this._secondLineTextProperties
    };
  }

  set announce(announce) {
    super._announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }
    return [this._FirstLine.announce, this._SecondLine.announce];
  }

  get _textW() {
    return this._focusW - (this.logo ? this.logoW + this.logoSpacing : 0);
  }

  get _focusW() {
    const scale =
      (this.hasFocus() ? this.focusScaleConst : this.unfocusScaleConst) || 1;
    return this.w * scale;
  }
}
