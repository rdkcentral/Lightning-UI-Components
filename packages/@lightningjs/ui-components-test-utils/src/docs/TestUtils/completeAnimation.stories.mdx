<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / completeAnimation" />

# completeAnimation

Returns a Promise that resolves once all animating properties have updated to their target value(s).

## Arguments

`completeAnimation(element, transitionProperties = [])`

| argument             | type              | default | description                                                 |
| -------------------- | ----------------- | ------- | ----------------------------------------------------------- |
| element              | Lightning.Element | -       | element with properties that transition value changes       |
| transitionProperties | array\|string     | []      | property name(s) that transition value changes on component |

<br />

##### Example: Testing element properties before and after their values have transitioned

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    return {
      TransitioningRect: {
        rect: true,
        color: 0xffffffff,
        w: 100,
        h: 100,
        x: 0,
        y: 0
      }
    };
  }

  _updatePosition() {
    this.tag('TransitioningRect').setSmooth('x', this.axisPosition);
    this.tag('TransitioningRect').setSmooth('y', this.axisPosition);
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this._updatePosition();
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. await the transitions of x and y values to finish before testing the result
it('should transition the x value of the component', async () => {
  const [example] = createExampleComponent();
  expect(example.tag('TransitioningRect').x).toBe(0);
  expect(example.tag('TransitioningRect').y).toBe(0);

  example.axisPosition = 50;
  await completeAnimation(example.tag('TransitioningRect'), ['x', 'y']);

  expect(example.tag('TransitioningRect').x).toBe(50);
  expect(example.tag('TransitioningRect').y).toBe(50);
});
```
