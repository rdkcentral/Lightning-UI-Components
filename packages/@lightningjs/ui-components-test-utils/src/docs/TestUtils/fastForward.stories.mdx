<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / fastForward" />

# fastForward

Force all running Lightning Transitions on one or more Lightning Elements to finish and update the transitioning property to its target value immediately.

This function is similar to [completeAnimation](/story/unit-testing-test-utils-completeanimation--page) in that it allows testing transitioning properties. Unlike `completeAnimation`, `fastForward` is not asynchronous, and does result in all running transitions to complete. If a test requires awaiting specific properties to finish transitioning, `completeAnimation` is the recommended utility function. The `fastForward` function forces transitions to finish, but does not wait for the template to update before proceeding through the test function. To ensure the template has updated following a call to `fastForward`, invoke [`testRenderer.update()`](/story/unit-testing-test-renderer-overview--page).
// TODO: add link directly to testRenderer.update() docs once those are written

## Arguments

`fastForward(elements)`

| argument | type                                           | default | description                                   |
| -------- | ---------------------------------------------- | ------- | --------------------------------------------- |
| elements | Array<Lightning.Element\> \| Lightning.Element | -       | element(s) with transitioning property values |

<br />

##### Example: Forcing running Transitions to finish before testing the result

```js
import lng from '@lightningjs/core';
import {
  makeCreateComponent,
  fastForward
} from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    const square = {
      rect: true,
      w: 100,
      h: 100,
      x: 0,
      y: 0,
      color: 0xeeeeeeee
    };
    return {
      TransitioningRectA: square,
      TransitioningRectB: square
    };
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this.tag('TransitioningRectA').setSmooth('x', value);
    this.tag('TransitioningRectB').setSmooth('y', value);
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. force all running transitions to complete, then test the result
it('should transition the x value of the component', () => {
  const [example, testRenderer] = createExampleComponent();
  expect(example.tag('TransitioningRectA').x).toBe(0);
  expect(example.tag('TransitioningRectB').y).toBe(0);

  //
  example.axisPosition = 50;
  const rectA = example.tag('TransitioningRectA');
  const rectB = example.tag('TransitioningRectB');
  fastForward([rectA, rectB]);
  testRenderer.update();

  expect(example.tag('TransitioningRectA').x).toBe(50);
  expect(example.tag('TransitioningRectB').y).toBe(50);
});
```
