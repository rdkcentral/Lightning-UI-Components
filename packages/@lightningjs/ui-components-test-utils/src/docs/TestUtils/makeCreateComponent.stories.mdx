<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Docs / Unit Testing / Test Utils / makeCreateComponent " />

# makeCreateComponent

Returns a function, referred to as a `createComponent` function, which is used to generate an instance of a component to run unit test cases against.

## Arguments

`makeCreateComponent(component, defaultConfig, defaultOptions)`

| argument       | type          | default                                  | description                                                                                                                |
| -------------- | ------------- | ---------------------------------------- | -------------------------------------------------------------------------------------------------------------------------- |
| component      | lng.Component | the component to generate an instance of | component to render an instance of                                                                                         |
| defaultConfig  | object        | {}                                       | properties that should be applied to the component by default                                                              |
| defaultOptions | object        | {}                                       | default options used when creating the component instance (these options are passed to the `TestRenderer.create` function) |

### `createComponent`

Returns an array of 2 elements:

1. an instance of the component to be tested against (generated by `testRenderer.getInstance()`)
2. a [testRenderer](/story/unit-testing-test-renderer-create--page#testrenderer-functions) object generated using the `component` argument and a merged object created from `defaultOptions` and `options`

#### Arguments

| argument | type   | default | description                                                                                                                                                                |
| -------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| config   | object | {}      | properties that should be applied to the component. This object is spread over the `defaultConfig` object.                                                                 |
| options  | object | {}      | options used when creating the component instance (these options are passed to the `TestRenderer.create` function). This object is spread over the `defaultOptions` object |

<br />

##### Example: Generating a `makeComponent` function and test component

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

// Lightning component to write unit tests for
class Example extends lng.Component {
  static _template() {
    return {
      Example: {
        text: { text: '' }
      }
    };
  }

  set exampleProp(value) {
    this._exampleProp = value;
    this.tag('Example').text.text = value;
  }

  get exampleProp() {
    return this._exampleProp;
  }
}

// unit tests for Example component

// generate a createComponent function to use in tests
const defaultConfig = { exampleProp: 'foo' };
const defaultOptions = { focused: false };
const createExampleComponent = makeCreateComponent(
  Example,
  defaultConfig,
  defaultOptions
);

// Ex. An instance of the component rendered with the properties defined in the defaultConfig and defaultOptions object
it('should render the exampleProp value as text', () => {
  const [example] = createExampleComponent();

  expect(example.exampleProp).toBe('foo');
  expect(example.tag('Example').text.text).toBe('foo');
  expect(example.hasFocus()).toBe(false);
});

// Ex. An instance of the component rendered with the properties defined in the config object
it('should support updating the text', () => {
  // overwrite the value of exampleProp defined in defaultConfig
  const [example] = createExampleComponent({
    exampleProp: 'bar'
  });

  expect(example.exampleProp).toBe('bar');
  expect(example.tag('Example').text.text).toBe('bar');
});

// Ex. An instance of the component rendered with options defined in the options object
it('should support rendering the component in a focused state', () => {
  // overwrite the value of focused defined in defaultOptions
  const [example] = createExampleComponent({}, { focused: true });

  expect(example.hasFocus()).toBe(true);
});

// Ex. Using a function provided by the testRenderer object (focus)
it('should allow focusing on the component', () => {
  const [example, testRenderer] = createExampleComponent();

  // focused was set to false in defaultOptions, so the componet will initially render without focus
  expect(example.hasFocus()).toBe(false);

  testRenderer.focus();

  expect(example.hasFocus()).toBe(true);
});
```

### spyOnMethods

A utility which enables running component methods asynchronously.

The `options` argument passed to the `createComponent` function accepts a field called `spyOnMethods`.
The value for `spyOnMethods` can be an array of method names to generate promises from. When one of those methods is invoked in the component it's associated promise will resolve, then the actual method will be invoked. The promises generated from the passed in methods are given the name `_${methodName}SpyPromise`.

<br />

##### Example: Spying on component methods and awaiting their execution before making assertions on their side effects

```js
class MethodSpiesExample extends Base {
  static get __componentName() {
    return 'MethodSpiesExample';
  }

  static _template() {
    return {
      Title: {
        type: TextBox
      },
      Subtitle: {
        type: TextBox,
        y: 100
      }
    };
  }

  static get tags() {
    return ['Title', 'Subtitle'];
  }

  static get properties() {
    return ['title', 'subtitle'];
  }

  _construct() {
    super._construct();
    this._title = 'title';
    this._subtitle = 'subtitle';
  }

  _update() {
    this._updateTitle();
    this._updateSubtitle();
  }

  _updateTitle() {
    this._Title.content = this.title;
  }
  _updateSubtitle() {
    this._Subtitle.content = this.subtitle;
  }
}

const createComponent = makeCreateComponent(MethodSpies);

it('should update the title and subtitle', async () => {
  const [methodSpiesComponent, testRenderer] = createComponent(
    {},
    { spyOnMethods: ['_update'] }
  );

  // wait for _update to be called on the initial render before proceeding
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe('title');
  expect(methodSpiesComponent._Subtitle.content).toBe('subtitle');

  const newTitle = 'new title';
  const newSubtitle = 'new subtitle';
  methodSpiesComponent.patch({
    title: newTitle,
    subtitle: newSubtitle
  });

  // wait for _update to be called in response to values in the properties array changing
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe(newTitle);
  expect(methodSpiesComponent._Subtitle.content).toBe(newSubtitle);
});
```

**Note: `spyOnMethods` can only be set in in the `options` object of `createComponent`. It can not be assigned to the `defaultOptions` object in `makeCreateComponent`**
