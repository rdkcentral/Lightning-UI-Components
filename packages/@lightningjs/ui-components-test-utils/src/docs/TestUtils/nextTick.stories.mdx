<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / nextTick" />

# nextTick

Creates a Promise that resolves after a defined amount of time. If no amount of time is specified, the Promise will resolve immediately.

## Arguments

`nextTick(wait = 0)`

| argument | type   | default | description                                                           |
| -------- | ------ | ------- | --------------------------------------------------------------------- |
| wait     | number | 0       | amount of time (in milliseconds) before the returned Promise resolves |
