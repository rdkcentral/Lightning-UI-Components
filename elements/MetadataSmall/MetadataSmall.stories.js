import lng from '@lightningjs/core';

import MetadataSmall from '.';
import mdx from './MetadataSmall.mdx';
import lightningbolt from '../../assets/images/ic_lightning_white_32.png';
import circle from '../../assets/images/circle.svg';
import { getHexColor } from '../../Styles/Colors';

export default {
  title: 'Elements / MetadataSmall',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        MetadataSmall: {
          type: MetadataSmall,
          w: 400,
          h: 200,
          title: args.title,
          data: args.data,
          progress: args.progress,
          logo: circle
        }
      };
    }

    _getFocused() {
      return this.tag('MetadataSmall');
    }
  };
Basic.args = {
  title: 'Tile Title',
  data: [
    '94%',
    {
      icon: lightningbolt,
      color: getHexColor('00ff00'),
      title: 'Green Lightning Bolt'
    },
    '86%',
    {
      icon:
        'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
      title: 'Rotten Tomatoes rating'
    }
  ],
  progress: 0
};
Basic.parameters = {
  argActions: {}
};

export const ClippedText = args =>
  class ClippedText extends lng.Component {
    static _template() {
      return {
        MetadataSmall: {
          type: MetadataSmall,
          w: 400,
          title: args.title,
          data: args.data,
          progress: args.progress,
          logo: args.showLogo ? lightningbolt : undefined,
          logoWidth: 1920,
          logoHeight: 1080
        }
      };
    }

    _getFocused() {
      return this.tag('MetadataSmall');
    }
  };
ClippedText.args = {
  title: 'Tile Title',
  // data: ["an example of text that will fade out i swear to god it'll fade out"],
  data:
    'description askdljfh sadf as jfhk asf as lfa sfdj alsk fd kask jldfalks d a sd fa lsdkjf lakjs dfk sdalkf al f',
  progress: 0,
  showLogo: false
};
ClippedText.parameters = {
  argActions: {}
};
