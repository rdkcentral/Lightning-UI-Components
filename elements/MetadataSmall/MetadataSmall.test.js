import TestUtils from '../../test/lightning-test-utils';
import MetadataSmall from '.';
import lightningbolt from '../../assets/images/ic_lightning_white_32.png';

const createComponent = TestUtils.makeCreateComponent(MetadataSmall, {
  title: 'title',
  w: 410,
  w: 230
});

describe('MetadataSmall', () => {
  let component, testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createComponent();
  });
  afterEach(() => {
    component = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should have a lighter or darker data alpha depending on the content', () => {
    component.data = 'Trending';
    expect(component._dataAlpha).toBe(component.styles.data.alphaLight);
    component.data = ['array', 'data'];
    expect(component._dataAlpha).toBe(component.styles.data.alphaDark);
  });

  it("should wrap the title smaller if there's a logo present", () => {
    let [metadataWithLogo, testRenderer2] = TestUtils.makeCreateComponent(
      MetadataSmall,
      {
        logo: lightningbolt,
        logoWidth: 1920,
        logoHeight: 1080
      }
    )();
    const fullWrap = component._Title.wordWrapWidth;
    const shortWrap = metadataWithLogo._Title.wordWrapWidth;
    expect(shortWrap).toBeLessThan(fullWrap);

    metadataWithLogo = null;
    testRenderer2 = null;
  });

  it('should show a progress bar if a progress value is provided', () => {
    let [metadataWithProgress, testRenderer2] = TestUtils.makeCreateComponent(
      MetadataSmall,
      {
        progress: 0.2
      }
    )();

    expect(metadataWithProgress._ProgressBar).toBeDefined();

    metadataWithProgress = null;
    testRenderer2 = null;
  });
});
