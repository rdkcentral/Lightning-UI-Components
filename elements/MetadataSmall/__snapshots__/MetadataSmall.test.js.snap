// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataSmall renders 1`] = `
Object {
  "Component": Object {
    "Content": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "DataClipContainer": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Data": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Data",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": Array [
                "Data",
              ],
              "type": "InlineContent",
              "visible": true,
              "w": 0,
              "x": 8,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "DataClipContainer",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "DataClipContainer",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Logo": Object {
          "active": false,
          "alpha": 0,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": false,
          "flex": false,
          "flexItem": false,
          "h": 32,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Logo",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Logo",
          ],
          "type": "Icon",
          "visible": true,
          "w": 32,
          "x": 278,
          "y": 152,
          "zIndex": 0,
        },
        "ProgressBarWrapper": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ProgressBarWrapper",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "ProgressBarWrapper",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Title": Object {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Title",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Title",
          ],
          "texture": Object {
            "fontFace": "XfinityBrownBold",
            "fontSize": 28,
            "letterSpacing": -0.2,
            "lineHeight": 40,
            "maxLines": 2,
            "maxLinesSuffix": "...",
            "text": "title",
            "textColor": 4076863487,
            "type": "TextTexture",
            "verticalAlign": "middle",
            "wordWrapWidth": 342,
          },
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 8,
          "y": 16,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 200,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Content",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Content",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 56,
      "zIndex": 0,
    },
  },
}
`;
